### PR TITLE
Add DisableInitContainer field in the CR

### PR DESCRIPTION
**Pull request by dkwon17**
### What does this PR do?
This PR adds the DisableInitContainer field in the CheCluster CR.

This field sets the `config.workspaces.persistUserHome.disableInitContainer` field in the Che-owned DWOC.

### Screenshot/screencast of this PR

image

### How to test this PR?

1. Prepare a patch file:
```bash
cat > /tmp/cr-patch.yaml <<EOF
apiVersion: org.eclipse.che/v2
kind: CheCluster
spec:
  devEnvironments:
    persistUserHome:
      enabled: true
      disableInitContainer: true
EOF
```

2. Deploy the operator:
 ```bash
 
## Openshift
./build/scripts/olm/test-catalog-from-sources.sh --cr-patch-yaml /tmp/cr-patch.yaml

## Minikube
./build/scripts/minikube-tests/test-operator-from-sources.sh --cr-patch-yaml /tmp/cr-patch.yaml
```

3. Verify that the `devworkspace-config` DWOC in the `eclipse-che` namespace has the following fields:
```
config:
  workspace:
    persistUserHome:
      enabled: true
      disableInitContainer: true
```

4. In the Che CR, `spec.devEnvironments.persistUserHome.disableInitContainer` to `false`. Verify that the DWOC has the following fields:
```
config:
  workspace:
    persistUserHome:
      enabled: true
      disableInitContainer: true
```

5.  In the Che CR, remove the `spec.devEnvironments.persistUserHome.disableInitContainer` field. Verify that the `disableInitContainer` has been removed from the DWOC:
```
config:
  workspace:
    persistUserHome:
      enabled: true
      disableInitContainer: true
```

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Development resource are up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-development-resources)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: api/v2/checluster_types.go [hunk 1]
```diff
@@ -479,6 +479,12 @@ type PersistentHomeConfig struct {
 	// Must be used with the 'per-user' or 'per-workspace' PVC strategy in order to take effect.
 	// Disabled by default.
 	Enabled *bool `json:"enabled,omitempty"`
+	// Determines whether the init container that initializes the persistent home directory should be disabled.
+	// When the `/home/user` directory is persisted, the init container is used to initialize the directory before
+	// the workspace starts. If set to true, the init container will not be created.
+	// This field is not used if the `workspace.persistUserHome.enabled` field is set to false.
```
Comment (AObuchow) [inline api/v2/checluster_types.go:485]: This sentence needs to be adapted for the Che Cluster CR: `workspace.persistUserHome.enabled` -> `devEnvironments.persistUserHome.enabled`

File: api/v2/checluster_types.go [hunk 2]
```diff
@@ -479,6 +479,12 @@ type PersistentHomeConfig struct {
...
+	// Enabled by default.
```
Comment (AObuchow) [inline api/v2/checluster_types.go:486]: I'm a bit confused, I thought this was [disabled by default](https://github.com/devfile/devworkspace-operator/pull/1251/files#diff-2658370b0497fa11f0682953637e6e2d237a568851f410b6de8e1ad202045a6cR48)? Do we have a docs bug in DWO?

**Review by AObuchow — APPROVED** [review]
Small documentation comments but overall LGTM

File: api/v2/checluster_types.go (hunk 1 above)
Comment (dkwon17) [reply api/v2/checluster_types.go]: Fixed

File: api/v2/checluster_types.go (hunk 2 above)
Comment (dkwon17) [reply api/v2/checluster_types.go:486]: Right, the value is set to false by default, but the init container itself enabled by default

Comment (AObuchow) [reply api/v2/checluster_types.go:486]: Ah okay that makes more sense. Maybe we should write "Set to false by default" or "The init container is enabled by default"? I'm okay with leaving it as is, but it almost reads as "DisableInitContainer is enabled by default = the init contianer is disabled by default"

**Review by ibuziuk — COMMENTED** [review]
@dkwon17 could you clarify what is the motivation and use-case behind this feature?
from the first glance `DisableInitContainer` looks like internal impl details rather than smth. to be exposed for configuration

**Review by tolusha — APPROVED** [review]

**Comment by dkwon17** [comment]
@ibuziuk The new field allows persistenthome setup from the tooling container's entrypoint instead of the initcontainer, since disabling the initcontainer means that persistent home can now only be set up by the tooling container (which was the beheaviour before this PR https://github.com/devfile/devworkspace-operator/pull/1251/files#diff-f1f3498fd50bbc235d4d09f5f5485a91f5a4a5cfeb20b127c96fc849a30ba56b).

File: api/v2/checluster_types.go [hunk 3]
```diff
@@ -479,6 +479,12 @@ type PersistentHomeConfig struct {
 	// Must be used with the 'per-user' or 'per-workspace' PVC strategy in order to take effect.
 	// Disabled by default.
 	Enabled *bool `json:"enabled,omitempty"`
+	// Determines whether the init container that initializes the persistent home directory should be disabled.
+	// When the `/home/user` directory is persisted, the init container is used to initialize the directory before
+	// the workspace starts. If set to true, the init container will not be created.
+	// This field is not used if the `devEnvironments.persistUserHome.enabled` field is set to false.
+	// Enabled by default.
+	DisableInitContainer *bool `json:"disableInitContainer,omitempty"`
```
[the remark this answers is not in the document]
Comment (dkwon17) [reply api/v2/checluster_types.go:487]: How about:
```
	// Determines whether the init container that initializes the persistent home directory should be disabled.
	// When the `/home/user` directory is persisted, the init container is used to initialize the directory before
	// the workspace starts. If set to true, the init container will not be created.
	// Disabling the init container allows persistent home to be prepared from the first workspace container's entrypoint script.
	// This field is not used if the `workspace.persistUserHome.enabled` field is set to false.
	// The init container is enabled by default.
```

WDYT @ibuziuk @AObuchow ?

[the remark this answers is not in the document]
Comment (AObuchow) [reply api/v2/checluster_types.go:487]: Looks good to me, though I'd modify the way "first workspace container" is said:
`// Disabling the init container allows persistent home to be prepared from the first workspace container's entrypoint script.`

I'm not sure what's the "standard" way of calling the main/tooling container? Do we say "tooling container" or "first container component"? 

Some alternatives:
- `Disabling the init container allows home persistence to be initialized by the entrypoint present in the workspace's first container component.`
- `Disabling the init container leaves the responsibility of initializing home persistence to the workspace's tooling container component, through an entrypoint, for example.`

[the remark this answers is not in the document]
Comment (dkwon17) [reply api/v2/checluster_types.go:488]: Thanks, in my latest commit I went with:
https://github.com/eclipse-che/che-operator/blob/11ddb357b9ee7c68cb51992631bc1c8499f7dc21/api/v2/checluster_types.go#L482-L488

**Comment by dkwon17** [comment]
I've made the changes to the description, could you please take a look again @ibuziuk

**Review by ibuziuk — APPROVED** [review]

**Pull request merged**